src/storage/src/render/upsert/types.rs

```diff
@@ -7,85 +7,478 @@
+}
+
+pub type UpsertValueAndSize = mz_rocksdb::GetResult<StateValue>;
+
+#[derive(Clone, Debug)]
+pub struct PutValue {
+    pub value: Option<UpsertValue>,
+    pub previous_persisted_size: Option<i64>,
+}
+
+/// In any `UpsertState` implementation, we need to support 2 modes:
+/// - Normal operation
+/// - Consolidation of snapshots (during rehydration).
+///
+/// This struct (and `Snapshotting`) is effectively a helper to simplify the logic that
+/// individual `UpsertState` implementations need to do to manage these 2 modes.
+///
+/// Normal operation is easy, we just store an ordinary `UpsertValue`, and allow the implementer
+/// to store it any way they want. During consolidation of snapshots, the logic is more complex.
+/// See the docs on `StateValue::merge_update` for more information.
+///
+/// This struct is not part of the `UpsertState` public API, but implementing that API without
+/// using it is considered hard-mode.
+#[derive(Clone, serde::Serialize, serde::Deserialize, Debug)]
+pub enum StateValue {
+    Snapshotting(Snapshotting),
+    Decoded(UpsertValue),
+}
+
+/// A value as produced during consolidation of a snapshot.
+#[derive(Clone, Default, serde::Serialize, serde::Deserialize, Debug)]
+pub struct Snapshotting {
+    value_xor: Vec<u8>,
+    len_sum: Wrapping<i64>,
+    checksum_sum: Wrapping<i64>,
+    diff_sum: Wrapping<i64>,
+}
+
+impl From<UpsertValue> for StateValue {
+    fn from(uv: UpsertValue) -> Self {
+        Self::Decoded(uv)
+    }
+}
+
+impl StateValue {
+    /// We use a XOR trick in order to accumulate the snapshot without having to store the full
+    /// unconsolidated history in memory. For all (value, diff) updates of a key we track:
+    /// - diff_sum = SUM(diff)
+    /// - checksum_sum = SUM(checksum(bincode(value)) * diff)
+    /// - len_sum = SUM(len(bincode(value)) * diff)
+    /// - value_xor = XOR(bincode(value))
+    ///
+    /// ## Correctness
+    ///
+    /// The method is correct because a well formed upsert snapshot will have for each key:
+    /// - Zero or one updates of the form (cur_value, +1)
+    /// - Zero or more pairs of updates of the form (prev_value, +1), (prev_value, -1)
+    ///
+    /// We are interested in extracting the cur_value of each key and discard all prev_values
+    /// that might be included in the stream. Since the history of prev_values always comes in
+    /// pairs, computing the XOR of those is always going to cancel their effects out. Also,
+    /// since XOR is commutative this property is true independent of the order. The same is
+    /// true for the summations of the length and checksum since the sum will contain the
+    /// unrelated values zero times.
+    ///
+    /// Therefore the accumulators will end up precisely in one of two states:
+    /// 1. diff == 0, checksum == 0, value == [0..] => the key is not present
+    /// 2. diff == 1, checksum == checksum(cur_value) value == cur_value => the key is present
```

Comment:
nit: Can a comment be added for the expected return value? That it will return `false` if a key was effectively added and `true` otherwise?

Reply:
Ditto, I wondered the same. The return value seems to indicate if the value if effectively deleted after the merge operation

Reply:
updated!